oshmem/mca/spml/ucx/spml_ucx.h

```diff
@@ -153,17 +153,52 @@ extern int mca_spml_ucx_fence(shmem_ctx_t ctx);
 extern int mca_spml_ucx_quiet(shmem_ctx_t ctx);
 extern int spml_ucx_progress(void);
 
+static void mca_spml_ucx_cache_mkey(mca_spml_ucx_ctx_t *ucx_ctx, sshmem_mkey_t *mkey, uint32_t segno, int dst_pe)
+{
+    ucp_peer_t *peer;
+
+    peer = &(ucx_ctx->ucp_peers[dst_pe]);
+    mkey_segment_init(&peer->mkeys[segno].super, mkey, segno);
+}
 
 static inline spml_ucx_mkey_t * 
-mca_spml_ucx_get_mkey(mca_spml_ucx_ctx_t *ucx_ctx, int pe, void *va, void **rva, mca_spml_ucx_t* module)
+mca_spml_ucx_get_mkey(shmem_ctx_t ctx, int pe, void *va, void **rva, mca_spml_ucx_t* module)
 {
     spml_ucx_cached_mkey_t *mkey;
+    mca_spml_ucx_ctx_t *ucx_ctx = (mca_spml_ucx_ctx_t *)ctx;
 
     mkey = ucx_ctx->ucp_peers[pe].mkeys;
     mkey = (spml_ucx_cached_mkey_t *)map_segment_find_va(&mkey->super.super, sizeof(*mkey), va);
     if (OPAL_UNLIKELY(NULL == mkey)) {
-        assert(module->get_mkey_slow);
-        return module->get_mkey_slow(pe, va, rva);
+        if (ucx_ctx != &mca_spml_ucx_ctx_default && pe == oshmem_my_proc_id()) {
+            mkey = mca_spml_ucx_ctx_default.ucp_peers[pe].mkeys;
+            mkey = (spml_ucx_cached_mkey_t *)map_segment_find_va(&mkey->super.super, sizeof(*mkey), va);
+            if (OPAL_UNLIKELY(NULL == mkey)) {
+                assert(module->get_mkey_slow);
+                return module->get_mkey_slow(ctx, pe, va, rva);
+            } else {
+                uint32_t segno = memheap_find_segnum(va);
+                sshmem_mkey_t *new_mkey = (sshmem_mkey_t *)calloc(1, sizeof(*new_mkey));
+                spml_ucx_mkey_t *new_ucx_mkey = &(ucx_ctx->ucp_peers[pe].mkeys[segno].key);
+                size_t len;
+
+                new_mkey->spml_context = new_ucx_mkey;
+
+                ucp_rkey_pack(mca_spml_ucx.ucp_context, mkey->key.mem_h, 
+                              &(new_mkey->u.data), &len);
+                
+                ucp_ep_rkey_unpack(ucx_ctx->ucp_peers[pe].ucp_conn,
+                                   new_mkey->u.data, &new_ucx_mkey->rkey);
+                new_mkey->len = len;
+                new_mkey->va_base = va;
+
+                *rva = map_segment_va2rva(&mkey->super, va);
+                return new_ucx_mkey;
+            }
+        } else {
+            assert(module->get_mkey_slow);
+            return module->get_mkey_slow(ctx, pe, va, rva);
```

Comment:
probably it will not request a new key from the remote, but will return the one cached for default context (if it's not a new mem region)